fix(upgrade): simplify registry URL resolution

PaulBratslavsky opened the pull request:
<!--
Hello 👋 Thank you for submitting a pull request.

To help us merge your PR, make sure to follow the instructions below:

- Create or update the tests
- Create or update the documentation at https://github.com/strapi/documentation
- Refer to the issue you are closing in the PR description: Fix #issue
- Specify if the PR is ready to be merged or work in progress (by opening a draft PR)

Please ensure you read the Contributing Guide: https://github.com/strapi/strapi/blob/main/CONTRIBUTING.md
-->

### What does it do?

Fix: Failed to parse URL from undefined/@strapi/strapi

### Why is it needed?

## The Problem

When running `npx @strapi/upgrade latest` on projects using **Yarn Classic (v1)**, users encountered this error:

```
[ERROR] Failed to parse URL from undefined/@strapi/strapi
```

## Root Cause

The upgrade tool queries the package manager for its configured NPM registry URL. The code was using:

```typescript
const registryCommands = {
  yarn: ['config', 'get', 'npmRegistryServer'],  // Wrong for Yarn v1!
  npm: ['config', 'get', 'registry'],
};
```

**The issue**: Yarn Classic (v1) and Yarn Berry (v2+) use different config keys:

| Yarn Version | Config Key | Command |
|--------------|------------|---------|
| Yarn Classic (v1) | `registry` | `yarn config get registry` |
| Yarn Berry (v2+) | `npmRegistryServer` | `yarn config get npmRegistryServer` |

When Yarn Classic is asked for `npmRegistryServer` (which doesn't exist), it returns the **literal string `"undefined"`** - not an empty string, not an error, but the actual word "undefined".

This caused the registry URL to become:
```
undefined/@strapi/strapi
```

Which is not a valid URL, causing the fetch to fail.

### How to test it?

Provide information about the environment and the path to verify the behaviour.

```bash
yarn test:unit
```

Tests cover:
- Custom registry via `NPM_REGISTRY_URL` environment variable
- Default registry fallback
- Trailing slash handling in registry URLs


### Related issue(s)/PR(s)

Let us know if this is related to any issue/pull request

https://github.com/strapi/strapi/issues/25014

PaulBratslavsky committed:
fix(upgrade): simplify registry URL resolution

innerdvations commented on packages/utils/upgrade/src/modules/npm/package.ts:
```diff
@@ -64,44 +62,12 @@ export class Package implements PackageInterface {
-    return constants.NPM_REGISTRY_URL.replace(/\/$/, '');
+  private getRegistryUrl(): string {
+    // Use NPM_REGISTRY_URL env var if set, otherwise use default registry
+    // This avoids spawning external processes (execa) to query package manager config
```
We want to maintain the original detection; we should be able to add your fix as an additional fallback in the original code though so that we support more package managers

PaulBratslavsky replied:
Thank you for the feedback.   🙂

PaulBratslavsky committed:
fix(upgrade): handle Yarn Classic registry config

   - Add yarn version detection to use correct config key
   - Yarn v1 uses 'registry', Yarn v2+ uses 'npmRegistryServer'
   - Guard against literal 'undefined' string from Yarn v1
   - Add tests for Yarn v1 and v2+ registry detection

innerdvations committed:
Merge branch 'develop' into fix/yarn-registry-url

innerdvations committed:
fix(upgrade): extract normalizeRegistryOutput helper

Move registry stdout validation into a shared helper and tighten registry
URL determination tests.

innerdvations committed:
Merge branch 'develop' into fix/yarn-registry-url
